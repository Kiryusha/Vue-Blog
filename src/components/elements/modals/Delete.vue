<template lang="pug">
  modal(
    name="delete"
    width="100%"
    height="auto"
    @before-open="beforeOpen"
  )
    ModalWrapper._delete
      .content
        .title Вы уверены, что хотите удалить новость?
        .close
          Button(
            @click="submitDeletion"
          ) Удалить
          Button(
            @click="$modal.hide('delete')"
          ) Отмена
</template>

<script>
import { mapState, mapMutations, mapActions } from 'vuex';
import ModalWrapper from 'Components/elements/modals/Modal';
import Button from 'Components/elements/general/Button';

export default {
  components: {
    ModalWrapper,
    Button,
  },
  data() {
    return {
      code: null,
    };
  },
  computed: {
    ...mapState({
      userId: state => state.user.id,
    }),
  },
  methods: {
    ...mapMutations(['DELETED_POST_FROM_LIST']),
    ...mapActions([
      'fetchCategories',
      'deletePost',
    ]),
    beforeOpen(event) {
      this.code = event.params.code;
    },
    async submitDeletion() {
      this.$modal.hide('delete');

      await this.deletePost({
        code: this.code,
        userId: this.userId,
      });

      this.DELETED_POST_FROM_LIST(this.code);

      this.fetchCategories();
    },
  },
};
</script>
