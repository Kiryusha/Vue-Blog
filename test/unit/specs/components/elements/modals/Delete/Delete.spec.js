// Delete.vue

import { shallowMount, createLocalVue } from '@vue/test-utils';
import Vuex from 'vuex';
import Delete from 'Components/elements/modals/Delete.vue';

const localVue = createLocalVue();
localVue.use(Vuex);

describe('Delete.vue', () => {
  const fetchCategories = jest.fn();
  const deletePost = jest.fn();
  const DELETED_POST_FROM_LIST = jest.fn();
  const config = {
    localVue,
    stubs: ['modal'],
    mocks: {
      $modal: {
        hide: jest.fn(),
      },
    },
  };

  let storeConfig = {
    state: {
      user: {
        id: '1a',
      },
    },
    mutations: {
      DELETED_POST_FROM_LIST: jest.fn(),
    },
    actions: {
      fetchCategories: () => {},
      deletePost: () => {},
    },
  };
  let store = new Vuex.Store(storeConfig);

  describe('component shallow rendering', () => {
    it('should match a snapshot', () => {
      const wrapper = shallowMount(Delete, { ...config, store });

      expect(wrapper.html()).toMatchSnapshot();
    });
  });

  describe('custom actions', () => {
    it('should set code data property on beforeOpen()', () => {
      const wrapper = shallowMount(Delete, { ...config, store });

      wrapper.vm.beforeOpen({
        params: {
          code: 'test',
        }
      })

      expect(wrapper.vm.code).toBe('test');
    });

    it('should call all delete functions on submitDeletion()', async () => {
      const wrapper = shallowMount(Delete, { ...config, store, methods: {
        deletePost,
        fetchCategories,
        DELETED_POST_FROM_LIST,
      }});

      wrapper.vm.beforeOpen({
        params: {
          code: 'test',
        }
      })

      await wrapper.vm.submitDeletion()

      expect(deletePost).toHaveBeenCalledTimes(1);
      expect(deletePost).toHaveBeenCalledWith({
        code: 'test',
        userId: '1a',
      });
      expect(fetchCategories).toHaveBeenCalledTimes(1);
      expect(DELETED_POST_FROM_LIST).toHaveBeenCalledTimes(1);
      expect(DELETED_POST_FROM_LIST).toHaveBeenCalledWith('test');
    });
  });
});
