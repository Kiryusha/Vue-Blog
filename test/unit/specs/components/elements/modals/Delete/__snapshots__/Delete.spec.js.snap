// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Delete.vue component shallow rendering should match a snapshot 1`] = `<modal-stub name="delete" width="100%" height="auto"></modal-stub>`;
